### PR TITLE
Add package zfit

### PR DESCRIPTION
**Pull request by jonas-eschle**
As requested by users, adding zfit to the spack repo.

Needs https://github.com/spack/spack/pull/42666, https://github.com/spack/spack/pull/42665

### CODE REVIEW

**Comment by jonas-eschle** [comment]
@spackbot help

**Comment by jonas-eschle** [comment]
@spackbot fix style

**Comment by jonas-eschle** [comment]
@adamjstewart tagging for review, as you've just merged similar libraries, you probably have a good expertise in this area

File: var/spack/repos/builtin/packages/py-zfit/package.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    # version("0.18.0", sha256="21d9479480f74945c67707b715780693bd4e94062c551bf41fe04a2eddb47fab")
+    version("0.17.0", sha256="cd60dfc360c82666af4e8dddd78edb0ab95a095b9dd0868457f0981dc03afa5a")
+    version("0.16.0", sha256="b3b170af23b61d7e265d6fb1bab1d052003f3fb41b3c537527cc1e5a1066dc10")
+    version("0.15.5", sha256="00a1138429e8a7f830c9e229b9c0bcd6071b95dadd8c87eb81191079fb679225")
+    version("0.14.1", sha256="66d1e349403f1d6c6350138d0f2b422046bcbdfb34fd95453dadae29a8b0c98a")
+
+    depends_on("python@3.9:", type=("build", "run"))
+    depends_on("py-setuptools", type="build")
+
+    runtype = (
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-zfit/package.py]: Much easier to use:
```python
with default_args(type=("build", "run")):
    depends_on("py-tensorflow@2.13", when="@0.15:0.17")
    ...
```

Comment (jonas-eschle) [reply var/spack/repos/builtin/packages/py-zfit/package.py]: Neat, refactored, thanks!

**Comment by jonas-eschle** [comment]
Ready from my side, thanks for all the hard work your putting in with merging and maintaining @adamjstewart , it's highly appreciated!

File: var/spack/repos/builtin/packages/py-zfit/package.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+    depends_on("python@3.9:", type=("build", "run"))
+    depends_on("py-setuptools", type="build")
+
+    variant("nlopt", default=False, description="Enable nlopt support")
+    variant("hs3", default=True, description="Enable serialization support")
+
+    with default_args(
+        type=("build", "run")
+    ):  # TODO: remove "build" once fixed in spack that tests need "run", not "build"
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-zfit/package.py]: If you put this comment on a different line you won't have to wrap the code on 3 lines

File: var/spack/repos/builtin/packages/py-zfit/package.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    # TODO: reactivate once TF 2.15 is ready https://github.com/spack/spack/pull/41069
+    # version("0.18.1", sha256="fbc6b3a636d8dc74fb2e69dfec5855f534c4583ec18efac9e9107ad45b18eb43")
+    # version("0.18.0", sha256="21d9479480f74945c67707b715780693bd4e94062c551bf41fe04a2eddb47fab")
+    version("0.17.0", sha256="cd60dfc360c82666af4e8dddd78edb0ab95a095b9dd0868457f0981dc03afa5a")
+    version("0.16.0", sha256="b3b170af23b61d7e265d6fb1bab1d052003f3fb41b3c537527cc1e5a1066dc10")
+    version("0.15.5", sha256="00a1138429e8a7f830c9e229b9c0bcd6071b95dadd8c87eb81191079fb679225")
+    version("0.14.1", sha256="66d1e349403f1d6c6350138d0f2b422046bcbdfb34fd95453dadae29a8b0c98a")
+
+    depends_on("python@3.9:", type=("build", "run"))
+    depends_on("py-setuptools", type="build")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-zfit/package.py]: Missing deps:
```toml
[build-system]
requires = [
    "setuptools>=42",
    "setuptools_scm[toml]>=3.4",
    "setuptools_scm_git_archive",
    "wheel"
]
```
You don't need wheel, it's already added by the base class.

File: var/spack/repos/builtin/packages/py-zfit/package.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+    # TODO: reactivate once TF 2.15 is ready https://github.com/spack/spack/pull/41069
+    # version("0.18.1", sha256="fbc6b3a636d8dc74fb2e69dfec5855f534c4583ec18efac9e9107ad45b18eb43")
+    # version("0.18.0", sha256="21d9479480f74945c67707b715780693bd4e94062c551bf41fe04a2eddb47fab")
+    version("0.17.0", sha256="cd60dfc360c82666af4e8dddd78edb0ab95a095b9dd0868457f0981dc03afa5a")
+    version("0.16.0", sha256="b3b170af23b61d7e265d6fb1bab1d052003f3fb41b3c537527cc1e5a1066dc10")
+    version("0.15.5", sha256="00a1138429e8a7f830c9e229b9c0bcd6071b95dadd8c87eb81191079fb679225")
+    version("0.14.1", sha256="66d1e349403f1d6c6350138d0f2b422046bcbdfb34fd95453dadae29a8b0c98a")
+
+    depends_on("python@3.9:", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-zfit/package.py]:
```suggestion
    depends_on("python@3.8:3.11", type=("build", "run"))
```

File: var/spack/repos/builtin/packages/py-zfit/package.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+        depends_on("py-dotmap")
+        depends_on("py-frozendict")
+        depends_on("py-hist")
+        depends_on("py-iminuit@2.3:")
+        depends_on("py-jacobi")
+        depends_on("py-numdifftools")
+        depends_on("py-numpy@1.16:")
+        depends_on("py-ordered-set")
+        depends_on("py-pandas")
+        depends_on("py-pydantic@1")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-zfit/package.py]:
```suggestion
        depends_on("py-pydantic@:1")
```

File: var/spack/repos/builtin/packages/py-zfit/package.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+        depends_on("py-tensorflow@2.0:2.12", when="@0.14")
+        depends_on("py-tensorflow-probability@0.20", when="@0.14")
+
+        with when("+nlopt"):
+            depends_on("nlopt@2.7.1: +python")
+
+        with when("+hs3"):
+            depends_on("py-asdf")
+
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-zfit/package.py:56]: Missing dep on py-attrs

File: var/spack/repos/builtin/packages/py-zfit/package.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+        depends_on("py-jacobi")
+        depends_on("py-numdifftools")
+        depends_on("py-numpy@1.16:")
+        depends_on("py-ordered-set")
+        depends_on("py-pandas")
+        depends_on("py-pydantic@1")
+        depends_on("py-pyyaml")
+        depends_on("py-scipy@1.2:")
+        depends_on("py-tabulate")
+        depends_on("py-texttable")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-zfit/package.py:78]: Missing py-typing-extensions

File: var/spack/repos/builtin/packages/py-zfit/package.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+
+        # TODO: reactivate once TF 2.15 is ready https://github.com/spack/spack/pull/41069
+        # depends_on("py-tensorflow@2.15", type=("run"), when="@0.18")
+        # depends_on("py-tensorflow-probability@0.23", type=("run"), when="@0.18")
+
+        depends_on("py-tensorflow@2.13", when="@0.15:0.17")
+        depends_on("py-tensorflow-probability@0.21", when="@0.15:0.17")
+
+        depends_on("py-tensorflow@2.12:2.13", when="@0.15:0.16")
+        depends_on("py-tensorflow-probability@0.20:0.21", when="@0.15:0.16")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-zfit/package.py:46]: 0.15 is covered in two separate sections. Which one is correct?

**Review by adamjstewart — CHANGES_REQUESTED** [review]

**Comment by jonas-eschle** [comment]
Fixed, refactored and styled, all should be fine now

File: var/spack/repos/builtin/packages/py-zfit/package.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+    # TODO: reactivate once TF 2.15 is ready https://github.com/spack/spack/pull/41069
+    # version("0.18.1", sha256="fbc6b3a636d8dc74fb2e69dfec5855f534c4583ec18efac9e9107ad45b18eb43")
+    # version("0.18.0", sha256="21d9479480f74945c67707b715780693bd4e94062c551bf41fe04a2eddb47fab")
+    version("0.17.0", sha256="cd60dfc360c82666af4e8dddd78edb0ab95a095b9dd0868457f0981dc03afa5a")
+    version("0.16.0", sha256="b3b170af23b61d7e265d6fb1bab1d052003f3fb41b3c537527cc1e5a1066dc10")
+    version("0.15.5", sha256="00a1138429e8a7f830c9e229b9c0bcd6071b95dadd8c87eb81191079fb679225")
+    version("0.14.1", sha256="66d1e349403f1d6c6350138d0f2b422046bcbdfb34fd95453dadae29a8b0c98a")
+
+    depends_on("python@3.8:3.11", type=("build", "run"))
+    depends_on("py-setuptools", type="build")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-zfit/package.py]:
```suggestion
    depends_on("py-setuptools@42:", type="build")
```

File: var/spack/repos/builtin/packages/py-zfit/package.py [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+    depends_on("py-setuptools-scm-git-archive", type="build")
+    depends_on("py-setuptools-scm@3.4:", type="build")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-zfit/package.py]:
```suggestion
    depends_on("py-setuptools-scm@3.4:+toml", type="build")
```

**Review by adamjstewart — CHANGES_REQUESTED** [review]

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**